Add manual links to controllers

Holzhaus opened the pull request.

Holzhaus committed:
controllers: Add support for manual links

Holzhaus committed:
controllers: Add manual pages to controller mappings

Holzhaus commented:
Requires mixxxdj/manual#122 to be merged first.

daschuer reviewed (changes requested):
Thank you. I have left some comments.

Unfortunately this is too big to review.
Can we revert the mass reformation?

Holzhaus committed:
controllers/controllerpreset: Use const reference

Holzhaus committed:
controllers/controllerpreset: Remove redunant inline keyword

Holzhaus commented:
> Unfortunately this is too big to review.

I removed the  commit. Btw, you can hide files by type in the "Files Changed" tab: https://github.com/mixxxdj/mixxx/pull/3108/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h

Holzhaus commented:
Moving the discussion from a removed (now orphaned) commit here for better visibility:

@daschuer 
> Can we auto generate a manual link?
> I can imagine that this will help to create a manual page.

@Holzhaus
> Can you explaint what I have in mind? The manual link is already partially auto-generated. You only need to specify the path, not the full URL.

@daschuer
> We may use the mapping file name or something..

@Holzhaus
> I don't think that's a good idea. Our controller mapping filenames have a very weird naming scheme (mixture of upper/lowercase chars, mixture of underscore and dashes, includes spaces). Some controller mappings include the manufacturers name, some don't. Sometimes the mapping creator appended his/her initials to the mapping filename and we failed to catch this during review.
> 
> Unfortunately, we can't change mapping names without breaking backwards compatibility (controller mapping would be unassigned).
>
> For the controller docs in the manual, I tried to use the name of mappings exactly as used by the manufacturer, including the company name and just amde them lowercase converted special chars and spaces into underscores.

@daschuer
> Here we are just talking about a default for new mappings.
> I think if the mapping file has a nice name, it is reasonable to generate the manual from it by default. This avoids namespace conflicts. If not we can specify any other name.
>
> Just an idea. if you still don't like it, it is OK for me.

Holzhaus commented:
> Here we are just talking about a default for new mappings.
> I think if the mapping file has a nice name, it is reasonable to generate the manual from it by default. This avoids namespace conflicts. If not we can specify any other name.

What do you mean by "namespace conflicts"? Can you give an example?

IMHO we could start estabilishing a new naming scheme for controller mappings to be able to do that. I'd appreciate that, but mazbe we should also postpone that until the new controller system is in place and keep the naming scheme for legacy mappings for consistency.

Also, I think we'd need another mechanism to distinguish user controller mappings from built-in ones. Right now, it's easy, because you just don't specify a manual page in custom mappings. If we auto-generate the link, we'd need to make an HTTP request to check if the page exists.

> I can imagine that this will help to create a manual page.

I  have doubts about this. For the Wiki page that would make sense, because you could just click the URL and be presented with an editing page. This is not the case for the Manual.

Holzhaus committed:
controllers: Update Forum links

Done using this python script:

    import re
    import requests
    import sys

    def replace_forum_url(matchobj):
        oldurl = matchobj.group(0).decode().replace("&amp;", "&")
        oldurl = "https://mixxx.discourse.group" + oldurl.partition("forums")[2]
        print(oldurl)
        req = requests.head(oldurl, allow_redirects=True)
        return req.url.replace("&", "&amp;").encode()

    for file in sys.argv[1:]:
        with open(file, mode="rb") as f:
            content = f.read()

        content = re.sub(rb'https?://mixxx.org/forums/[\w\?&;=\/\.]+', replace_forum_url, content)

        with open(file, mode="wb") as f:
            f.write(content)

Holzhaus committed:
controllers: Remove obsolete wiki links

daschuer commented on res/controllers/Vestax VCI-300.midi.xml:
```diff
@@ -4,7 +4,7 @@
 		<author>Uwe Klotz</author>
 		<forums>http://www.mixxx.org/forums/viewtopic.php?f=7&amp;t=4021</forums>
-		<wiki>http://mixxx.org/wiki/doku.php/vestax_vci-300</wiki>
+        <manual>/hardware/controllers/vestax_vci_300.html</manual>
```
Did you consider to remove the hardware/controllers part here?
This allows to only change a single place if this ever changes

Holzhaus replied:
Good point, I could also remove the file extension in case we ever switch to the HTML-folders sphinx builder.

Holzhaus replied:
Done.

Holzhaus committed:
controllers: Don't specify full path in <manual> tag

Holzhaus committed:
ControllerPreset: Rename manualPath to manualPage and add manualLink

daschuer commented:
Thank you. LGTM

the pre-commit hook reports some white space issues. 
Lets ignore them for now to not clutter this PR with unrelated changes.

Holzhaus commented:
> the pre-commit hook reports some white space issues.
> Lets ignore them for now to not clutter this PR with unrelated changes.

Yes, I had to skip a bunch of hooks because the formatting of our XML files is completely broken. This is why I opened #3107.
